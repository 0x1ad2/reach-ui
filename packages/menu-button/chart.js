"use strict";

exports.__esModule = true;
exports["default"] = void 0;

var _xstate = require("xstate");

function _extends() {
  _extends =
    Object.assign ||
    function(target) {
      for (var i = 1; i < arguments.length; i++) {
        var source = arguments[i];
        for (var key in source) {
          if (Object.prototype.hasOwnProperty.call(source, key)) {
            target[key] = source[key];
          }
        }
      }
      return target;
    };
  return _extends.apply(this, arguments);
}

var openEvents = {
  KEY_ESCAPE: {
    target: "idle",
    actions: ["focusButton"]
  },
  KEY_ARROW_UP: {
    target: "open:selecting",
    actions: ["navigatePrev"]
  },
  KEY_ARROW_DOWN: {
    target: "open:selecting",
    actions: ["navigateNext"]
  },
  BLUR: {
    target: "idle",
    cond: "clickedNonMenuItem"
  },
  ITEM_POINTER_ENTER: {
    target: "open:selecting",
    actions: ["highlightItem"]
  },
  ITEM_POINTER_LEAVE: {
    target: "open:selecting",
    actions: ["resetIndex"]
  },
  ITEM_POINTER_DOWN: "open:clickingItem",
  KEY_ENTER: {
    target: "open:confirming",
    cond: "hasHighlight"
  },
  KEY_SPACE: {
    target: "open:confirming",
    cond: "hasHighlight"
  },
  HOME: {
    target: "open:selecting",
    actions: ["highlightFirst"]
  },
  END: {
    target: "open:selecting",
    actions: ["highlightLast"]
  }
};
var chart = {
  id: "menu-button",
  context: {
    searchStartIndex: -1,
    search: "",
    activeIndex: -1,
    // barf
    button: null,
    items: null
  },
  initial: "idle",
  states: {
    "open:confirming": {
      entry: ["assignRefsForClose"],
      after: {
        200: {
          target: "idle",
          actions: ["focusButton", "selectItem"]
        }
      }
    },
    idle: {
      on: {
        // BUTTON_CLICK: {
        //   target: "open:selecting",
        //   actions: ["resetIndex"]
        // },
        BUTTON_POINTER_DOWN: {
          target: "open:clickingButton",
          actions: ["resetIndex"]
        },
        KEYDOWN_SPACE: {
          target: "open:selecting",
          actions: ["resetIndex", "focusMenu"]
        },
        KEYDOWN_ENTER: {
          target: "open:selecting",
          actions: ["resetIndex", "focusMenu"]
        },
        ARROW_DOWN: {
          target: "open:selecting",
          actions: ["highlightFirst"]
        },
        ARROW_UP: {
          target: "open:selecting",
          actions: ["highlightLast"]
        }
      }
    },
    "open:selecting": {
      entry: ["disableTooltips", "focusMenu"],
      exit: ["enableTooltips"],
      on: _extends({}, openEvents, {
        KEY_PRESS: {
          target: "open:searching",
          actions: [
            // "resetSearch"
            "concatSearch",
            "setSearchStartIndex"
          ]
        }
      })
    },
    "open:selectingWithDrag": {
      entry: ["disableTooltips"],
      exit: ["enableTooltips"],
      on: _extends({}, openEvents, {
        ITEM_POINTER_UP: {
          target: "open:confirming"
        },
        DOC_POINTER_UP: {
          target: "idle",
          actions: ["focusButton"]
        },
        BUTTON_POINTER_UP: {
          // blocks DOC_POINTER_UP
          target: "idle",
          actions: ["focusButton"]
        },
        ITEM_POINTER_ENTER: {
          target: "open:selectingWithDrag",
          actions: ["highlightItem"]
        },
        ITEM_POINTER_LEAVE: {
          target: "open:selectingWithDrag",
          actions: ["resetIndex"]
        }
      })
    },
    // TODO THIS IS ALL BUSTED
    "open:searching": {
      entry: ["highlightSearchMatch"],
      exit: ["resetSearchStartIndex"],
      after: {
        1000: {
          target: "open:selecting",
          actions: ["resetSearch"]
        }
      },
      on: _extends({}, openEvents, {
        KEY_PRESS: {
          target: "open:searching",
          actions: ["concatSearch"]
        }
      })
    },
    "open:clickingButton": {
      after: {
        500: "open:selectingWithDrag"
      },
      on: {
        BUTTON_POINTER_UP: "open:selecting",
        BUTTON_POINTER_LEAVE: "open:selectingWithDrag"
      }
    },
    "open:clickingItem": {
      after: {
        500: "open:selectingWithDrag"
      },
      on: {
        ITEM_POINTER_LEAVE: "open:selectingWithDrag",
        ITEM_POINTER_UP: {
          target: "open:confirming"
        }
      }
    }
  }
};
var actions = {
  focusButton: function focusButton(ctx, event) {
    ((event.refs && event.refs.button) || ctx.button).focus();
  },
  focusMenu: function focusMenu(ctx, event) {
    // need to let the keydown event finish before moving focus
    requestAnimationFrame(function() {
      event.refs.menu.focus();
    });
  },
  // index stuff
  highlightFirst: (0, _xstate.assign)({
    activeIndex: 0
  }),
  highlightLast: (0, _xstate.assign)({
    activeIndex: function activeIndex(ctx) {
      return ctx.items().length - 1;
    }
  }),
  highlightItem: (0, _xstate.assign)({
    activeIndex: function activeIndex(ctx, event) {
      return event.index;
    }
  }),
  navigateNext: (0, _xstate.assign)({
    activeIndex: function activeIndex(ctx, event) {
      var items = event.refs.items;
      console.log(items);
      return (ctx.activeIndex + 1) % items.length;
    }
  }),
  navigatePrev: (0, _xstate.assign)({
    activeIndex: function activeIndex(ctx, event) {
      var items = event.refs.items;
      return (ctx.activeIndex + items.length - 1) % items.length;
    }
  }),
  resetIndex: (0, _xstate.assign)({
    activeIndex: -1
  }),
  // tooltips
  disableTooltips: function disableTooltips() {
    window.__REACH_DISABLE_TOOLTIPS = false;
  },
  enableTooltips: function enableTooltips() {
    window.__REACH_DISABLE_TOOLTIPS = true;
  },
  // Search
  resetSearch: (0, _xstate.assign)({
    search: ""
  }),
  concatSearch: (0, _xstate.assign)({
    search: function search(ctx, event) {
      return ctx.search + event.key;
    }
  }),
  setSearchStartIndex: (0, _xstate.assign)({
    searchStartIndex: function searchStartIndex(ctx) {
      return ctx.activeIndex;
    }
  }),
  resetSearchStartIndex: (0, _xstate.assign)({
    searchStartIndex: -1
  }),
  highlightSearchMatch: (0, _xstate.assign)({
    activeIndex: function activeIndex(ctx, event) {
      var searchStartIndex = ctx.searchStartIndex,
        search = ctx.search;
      var searchString = search.toLowerCase();
      var items = event.refs.items;
      var reordered = items
        .slice(searchStartIndex + 1)
        .concat(items.slice(0, searchStartIndex));

      var _loop = function _loop(i, l) {
        var itemText = reordered[i].searchText.toLowerCase();

        if (itemText.startsWith(searchString)) {
          // adjust the index back since we rearranged them
          // there is a math way to do this like:
          // return searchStartIndex + 1 + i % items.length;
          // but it's too late right now
          return {
            v: items.findIndex(function(item) {
              return item === reordered[i];
            })
          };
        }
      };

      for (var i = 0, l = reordered.length; i < l; i++) {
        var _ret = _loop(i, l);

        if (typeof _ret === "object") return _ret.v;
      }

      return -1;
    }
  }),
  selectItem: function selectItem(ctx, event) {
    var items = ctx.items;
    items[ctx.activeIndex].onSelect();
  },
  assignRefsForClose: (0, _xstate.assign)({
    button: function button(ctx, event) {
      return event.refs.button;
    },
    items: function items(ctx, event) {
      return event.refs.items;
    }
  })
};
var guards = {
  hasHighlight: function hasHighlight(ctx) {
    return ctx.activeIndex > -1;
  },
  clickedNonMenuItem: function clickedNonMenuItem(ctx, event) {
    return !event.refs.menu.contains(event.relatedTarget);
  }
};

if (process.env.NODE_ENV !== "production") {
  validate(chart, actions, guards);
}

function validate(chart, actions, guards) {
  var usedActions = {};
  var usedGuards = {};

  for (var state in chart.states) {
    var eventActions = [];
    var entry = chart.states[state].entry;
    var exit = chart.states[state].exit;
    if (entry) eventActions.push.apply(eventActions, entry);
    if (exit) eventActions.push.apply(eventActions, exit);

    var events = _extends(
      {},
      chart.states[state].on,
      {},
      chart.states[state].after
    );

    for (var event in events) {
      if (events[event].actions) {
        eventActions.push.apply(eventActions, events[event].actions);
      }

      var guard = events[event].cond;

      if (guard) {
        usedGuards[guard] = true;

        if (!guards[guard]) {
          console.warn(
            'Guard not found: "' +
              guard +
              '" for ' +
              chart.id +
              ' "' +
              state +
              '"'
          );
        }
      }
    }

    for (
      var _i = 0, _eventActions = eventActions;
      _i < _eventActions.length;
      _i++
    ) {
      var action = _eventActions[_i];
      usedActions[action] = true;

      if (!actions[action]) {
        console.warn(
          'Action not found: "' +
            action +
            '" for ' +
            chart.id +
            ' "' +
            state +
            '"'
        );
      }
    }
  }

  for (var _action in actions) {
    if (!usedActions[_action]) {
      console.warn(
        'Defined action "' + _action + '" is not used in the chart.'
      );
    }
  } // for (let guard in guards) {
  //   if (!usedGuards[guard]) {
  //     console.warn(`Defined guard "${guard}" is not used in the chart.`);
  //   }
  // }
}

var _default = (0, _xstate.Machine)(chart, {
  actions: actions,
  guards: guards
});

exports["default"] = _default;
