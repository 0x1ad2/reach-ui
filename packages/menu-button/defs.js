"use strict";

exports.__esModule = true;
exports.item = exports.menu = exports.popover = exports.button = exports.container = void 0;

var _chart = _interopRequireDefault(require("./chart"));

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : { default: obj };
}

var container = {
  type: "ROOT_PROVIDER",
  displayName: "MenuButtonContainer",
  chart: _chart["default"],
  refs: {
    button: null,
    menu: null,
    items: []
  },
  setup: function setup(send) {
    var up = function up() {
      return send("DOC_POINTER_UP");
    };

    document.addEventListener("pointerup", up);
    return function() {
      return document.removeEventListener("pointerup", up);
    };
  }
};
exports.container = container;
var button = {
  displayName: "MenuButton",
  tagName: "button",
  ref: "button",
  events: {
    onClick: "BUTTON_CLICK",
    onPointerDown: leftClick("BUTTON_POINTER_DOWN"),
    onPointerUp: "BUTTON_POINTER_UP",
    onPointerMove: "BUTTON_POINTER_MOVE",
    onPointerLeave: "BUTTON_POINTER_LEAVE",
    onKeyDown: switchKey({
      " ": "KEYDOWN_SPACE",
      Enter: "KEYDOWN_ENTER",
      ArrowDown: "ARROW_DOWN",
      ArrowUp: "ARROW_UP"
    })
  },
  attrs: {
    type: "button",
    id: function id(state, ctx, rootId) {
      return rootId + "--button";
    },
    "aria-haspopup": "menu",
    "aria-expanded": function ariaExpanded(state) {
      return state.startsWith("open");
    },
    "aria-controls": function ariaControls(state, ctx, rootId) {
      // TODO: put rootId in ctx
      return rootId + "--menu";
    },
    "data-reach-menu-button": ""
  }
};
exports.button = button;
var popover = {
  type: "POPOVER",
  displayName: "MenuPopover",
  targetRef: "button",
  attrs: {
    "data-reach-menu-popover": ""
  }
};
exports.popover = popover;
var menu = {
  type: "INDEXED_PARENT",
  displayName: "Menu",
  tagName: "div",
  ref: "menu",
  descendants: "items",
  attrs: {
    role: "menu",
    id: function id(state, ctx, rootId) {
      return rootId + "--menu";
    },
    tabIndex: "-1",
    "aria-labelledby": function ariaLabelledby(state, ctx, rootId) {
      return rootId + "--button";
    },
    "aria-activedescendant": function ariaActivedescendant(state, ctx, rootId) {
      return rootId + "--menu-item-" + ctx.activeIndex;
    },
    "data-reach-menu": "",
    "data-closed": function dataClosed(state) {
      return !state.startsWith("open") ? "" : undefined;
    }
  },
  events: {
    onKeyDown: switchKey({
      Tab: function Tab(event) {
        return event.preventDefault();
      },
      Escape: "KEY_ESCAPE",
      ArrowDown: "KEY_ARROW_DOWN",
      ArrowUp: "KEY_ARROW_UP",
      Enter: "KEY_ENTER",
      " ": "KEY_SPACE",
      Home: "HOME",
      End: "END"
    }),
    onBlur: "BLUR",
    onPointerLeave: "POINTER_LEAVE_MENU",
    onKeyPress: function onKeyPress(event, send) {
      // TODO: just send the event all the time?
      send({
        type: "KEY_PRESS",
        key: event.key
      });
    }
  }
}; // YOU ARE HERE:
// Trying to figure out how to abstract these "descendants"

exports.menu = menu;
var item = {
  type: "INDEXED_CHILD",
  tagName: "div",
  events: {
    onPointerDown: "ITEM_POINTER_DOWN",
    onPointerUp: "ITEM_POINTER_UP",
    onPointerEnter: function onPointerEnter(event, send, index) {
      send({
        type: "ITEM_POINTER_ENTER",
        index: index
      });
    },
    onPointerLeave: "ITEM_POINTER_LEAVE"
  },
  attrs: {
    role: "menuitem",
    id: function id(state, ctx, rootId, index) {
      return rootId + "--menu-item-" + index;
    },
    "data-reach-menu-item": "",
    "data-selected": function dataSelected(state, ctx, rootId, index) {
      return ctx.activeIndex === index ? "" : undefined;
    },
    "data-confirming": function dataConfirming(state, ctx, rootId, index) {
      return ctx.activeIndex === index && state === "open:confirming"
        ? ""
        : undefined;
    }
  }
};
exports.item = item;

function switchKey(keys) {
  return function(event, send) {
    var handler = keys[event.key];

    if (handler) {
      if (typeof handler === "function") {
        handler(event, send);
      } else {
        send(handler);
      }
    }
  };
}

function leftClick(eventName) {
  return function(event, send) {
    if (event.button === 0) {
      send(eventName);
    }
  };
}
