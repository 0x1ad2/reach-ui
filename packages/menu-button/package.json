{
  "name": "@reach/menu-button",
  "version": "0.7.0",
  "description": "Accessible React button dropdown menu.",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/reach/reach-ui.git",
    "directory": "packages/menu-button"
  },
  "author": "React Training <hello@reacttraining.com>",
  "license": "MIT",
  "main": "dist/index.js",
  "umd:main": "dist/menu-button.umd.production.js",
  "module": "dist/menu-button.esm.js",
  "typings": "dist/index.d.ts",
  "files": [
    "dist",
    "styles.css"
  ],
  "scripts": {
    "build": "cross-env NODE_ENV=production tsdx build --format=cjs,esm,umd"
  },
  "dependencies": {
    "@reach/auto-id": "^0.7.0",
    "@reach/popover": "^0.7.0",
    "@reach/utils": "^0.7.1",
    "prop-types": "^15.7.2",
    "warning": "^4.0.3"
  },
  "peerDependencies": {
    "react": "^16.8.0",
    "react-dom": "^16.8.0"
  }
}
