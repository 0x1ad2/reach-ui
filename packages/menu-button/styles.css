/* Used to detect in JavaScript if apps have loaded styles or not. */
:root {
  --reach-menu-button: 1;
}

[data-reach-menu-popover] {
  font-family: sans-serif;
  display: block;
  position: absolute;
}

[data-reach-menu-list],
[data-reach-menu] {
  white-space: nowrap;
  border: solid 1px hsla(0, 0%, 0%, 0.25);
  background: hsla(0, 100%, 100%, 0.99);
  outline: none;
  padding: 0.5rem 0;
  font-size: 85%;
  user-select: none;
  visibility: visible;
  opacity: 1;
}

[data-reach-menu][data-closed] {
  visibility: hidden;
  opacity: 0;
  transition: visibility 0s linear 150ms, opacity 150ms ease;
}

[data-reach-menu-item] {
  display: block;
}

/*
 The dom structure of a MenuLink is reach-menu-item > a,
 so to target all items we can use `data-reach-menu-item`
*/
[data-reach-menu-item] {
  /*
    These are styled in one rule instead of something like a[data-reach-menu-item]
    and li[data-reach-menu-item] so that apps don't have to fight specificity and
    can style both li and a menu items with one rule,
    ie: `[data-selected] { background: red; }`.
    Otherwise they'd have to define two styles, one for a and one for li.
  */

  /* reach-menu-item */
  cursor: pointer;

  /* a */
  display: block;
  color: inherit;
  font: inherit;
  text-decoration: initial;

  /* both */
  padding: 5px 20px;
}

@keyframes flash {
  0% {
    background: rgb(83, 147, 248);
    color: white;
    opacity: 1;
  }
  50% {
    opacity: 0.5;
    background: inherit;
    color: inherit;
  }
  100% {
    background: rgb(83, 147, 248);
    color: white;
    opacity: 1;
  }
}

[data-reach-menu-item][data-selected] {
  background: rgb(79, 144, 248);
  color: white;
}

[data-reach-menu-item][data-selected][data-confirming] {
  animation: flash 100ms;
  animation-iteration-count: 1;
}
