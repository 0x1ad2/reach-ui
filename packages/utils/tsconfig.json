{
  "extends": "../../tsconfig.json",
  "include": ["src", "types", "../../types"],
  "compilerOptions": {
    "baseUrl": "./",
    "rootDir": "./",
    "paths": {
      "*": ["src/*", "node_modules/*"]
    }
  }
}
