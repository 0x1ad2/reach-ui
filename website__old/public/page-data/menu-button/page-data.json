{
  "componentChunkName": "component---src-pages-menu-button-mdx",
  "path": "/menu-button/",
  "result": {
    "pageContext": { "isCreatedByStatefulCreatePages": true, "frontmatter": {} }
  }
}
