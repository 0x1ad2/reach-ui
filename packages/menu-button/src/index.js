import adapt from "@reach/adapter";
import { container, button, popover, menu, item } from "./defs";

export const MenuContainer = adapt(container);
export const MenuButton = adapt(button);
export const MenuPopover = adapt(popover);
export const Menu = adapt(menu);
export const MenuItem = adapt(item, {
  registerDescendant: props => {
    return { name: props.children, onSelect: props.onSelect };
  }
});
