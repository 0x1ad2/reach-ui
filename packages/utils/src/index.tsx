/* eslint-disable no-unused-vars */

import {
  useRef,
  useMemo,
  useEffect,
  isValidElement,
  cloneElement
} from "react";
import {
  As,
  AssignableRef,
  ComponentWithAs,
  ComponentWithForwardedRef,
  PropsFromAs,
  PropsWithAs
} from "./types";
import React from "react";

let checkedPkgs: { [key: string]: boolean } = {};

/**
 * When in dev mode, checks that styles for a given @reach package are loaded.
 *
 * @param packageName Name of the package to check.
 * @example checkStyles("dialog") will check for styles for @reach/dialog
 */
// @ts-ignore
let checkStyles = (packageName: string): void => {};

if (__DEV__) {
  checkStyles = (pkg: string) => {
    console.warn({ env: process.env.NODE_ENV });

    // only check once per package
    if (checkedPkgs[pkg]) return;
    checkedPkgs[pkg] = true;

    if (
      process.env.NODE_ENV !== "test" &&
      parseInt(
        window
          .getComputedStyle(document.body)
          .getPropertyValue(`--reach-${pkg}`),
        10
      ) !== 1
    ) {
      console.warn(
        `@reach/${pkg} styles not found. If you are using a bundler like webpack or parcel include this in the entry file of your app before any of your own styles:

    import "@reach/${pkg}/styles.css";

  Otherwise you'll need to include them some other way:

    <link rel="stylesheet" type="text/css" href="node_modules/@reach/${pkg}/styles.css" />

  For more information visit https://ui.reach.tech/styling.
  `
      );
    }
  };
}

export { checkStyles };

/**
 * Passes or assigns an arbitrary value to a ref function or object.
 *
 * @param ref
 * @param value
 */
export function assignRef<T = any>(ref: AssignableRef<T>, value: any) {
  if (ref == null) return;
  if (typeof ref === "function") {
    ref(value);
  } else {
    try {
      // @ts-ignore
      ref.current = value;
    } catch (error) {
      throw new Error(`Cannot assign value "${value}" to ref "${ref}"`);
    }
  }
}

export function cloneValidElement<P>(
  element: React.ReactElement<P> | React.ReactNode,
  props?: Partial<P> & React.Attributes,
  ...children: React.ReactNode[]
): React.ReactElement<P> | React.ReactNode {
  if (!isValidElement(element)) {
    return element;
  }
  return cloneElement(element, props, ...children);
}

/**
 * Get the scrollbar offset distance.
 */
export function getScrollbarOffset() {
  try {
    if (window.innerWidth > document.documentElement.clientWidth) {
      return window.innerWidth - document.documentElement.clientWidth;
    }
  } catch (err) {}
  return 0;
}

/**
 * Joins strings to format IDs for compound components.
 *
 * @param args
 */
export function makeId(...args: (string | number)[]) {
  return args.join("--");
}

/**
 * No-op function.
 */
export function noop(): void {}

/**
 * Passes or assigns a value to multiple refs (typically a DOM node). Useful for
 * dealing with components that need an explicit ref for DOM calculations but
 * also forwards refs assigned by an app.
 *
 * @param refs Refs to fork
 */
export function useForkedRef<T = any>(...refs: AssignableRef<T>[]) {
  return useMemo(() => {
    if (refs.every(ref => ref == null)) {
      return null;
    }
    return (node: any) => {
      refs.forEach(ref => {
        assignRef(ref, node);
      });
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, refs);
}

/**
 * Returns the previous value of a reference after a component update.
 *
 * @param value
 */
export function usePrevious<T = any>(value: T) {
  const ref = useRef<T | null>(null);
  useEffect(() => {
    ref.current = value;
  }, [value]);
  return ref.current;
}

/**
 * Call an effect after a component update, skipping the initial mount.
 *
 * @param effect Effect to call
 * @param deps Effect dependency list
 */
export function useUpdateEffect(effect: () => any, deps?: any[]) {
  const mounted = useRef(false);
  useEffect(() => {
    if (mounted.current) {
      effect();
    } else {
      mounted.current = true;
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, deps);
}

/**
 * Wraps a lib-defined event handler and a user-defined event handler, returning
 * a single handler that allows a user to prevent lib-defined handlers from
 * firing.
 *
 * @param theirHandler User-supplied event handler
 * @param ourHandler Library-supplied event handler
 */
export function wrapEvent<E extends React.SyntheticEvent | Event>(
  theirHandler: ((event: E) => any) | undefined,
  ourHandler: (event: E) => any
): (event: E) => any {
  return event => {
    theirHandler && theirHandler(event);
    if (!event.defaultPrevented) {
      return ourHandler(event);
    }
  };
}

/**
 * This is a hack to stop TS errors from dynamic components with an `as` prop
 * TODO: Eventually we should probably just try to get the type defs above
 * working across the board, but ain't nobody got time for that mess!
 *
 * @param Comp
 */
export function forwardRefWithAs<T extends As, P>(
  Comp: (props: PropsFromAs<T, P>, ref: React.RefObject<any>) => JSX.Element
) {
  return React.forwardRef(Comp as any) as ComponentWithAs<T, P>;
}

export {
  As,
  AssignableRef,
  ComponentWithAs,
  ComponentWithForwardedRef,
  PropsFromAs,
  PropsWithAs
};
