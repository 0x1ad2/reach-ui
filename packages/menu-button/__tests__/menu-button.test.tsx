import React from "react";
import { render, act, withMarkup, fireEvent } from "$test/utils";
import { axe } from "jest-axe";
import { Menu, MenuList, MenuButton, MenuItem } from "@reach/menu-button";

describe("<MenuButton />", () => {
  it("should not have basic a11y issues", async () => {
    let { container } = render(<BasicMenuButton />);
    expect(await axe(container)).toHaveNoViolations();
  });

  it("should toggle on button click", () => {
    let { getByRole, getByText, queryByText } = render(<BasicMenuButton />);
    let getByTextWithMarkup = withMarkup(getByText);
    let queryByTextWithMarkup = withMarkup(queryByText);

    // Menu opens on mousedown, not click!
    function clickButton() {
      fireEvent.mouseDown(getByRole("button"));
      fireEvent.mouseUp(getByRole("button"));
    }

    expect(queryByTextWithMarkup("Create a Copy")).not.toBeTruthy();
    act(clickButton);
    expect(getByTextWithMarkup("Create a Copy")).toBeInTheDocument();
  });
});

function BasicMenuButton() {
  return (
    <Menu>
      <MenuButton id="example-button">
        Actions <span aria-hidden="true">â–¾</span>
      </MenuButton>
      <MenuList>
        <input type="text" />
        <MenuItem onSelect={() => jest.fn()}>Download</MenuItem>
        <MenuItem onSelect={() => jest.fn()}>Create a Copy</MenuItem>
        <MenuItem onSelect={() => jest.fn()}>Mark as Draft</MenuItem>
        <MenuItem onSelect={() => jest.fn()}>Delete</MenuItem>
      </MenuList>
    </Menu>
  );
}
