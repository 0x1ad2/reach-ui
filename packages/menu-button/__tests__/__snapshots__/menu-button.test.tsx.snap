// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuButton /> should match the snapshot 1`] = `
<DocumentFragment>
  <button
    aria-controls="menu--REACH-ID"
    aria-haspopup="true"
    data-reach-menu-button=""
    id="example-button"
    type="button"
  >
    Actions 
    <span
      aria-hidden="true"
    >
      â–¾
    </span>
  </button>
</DocumentFragment>
`;
